### PR TITLE
Re-use deleted node IDs in incremental mode

### PR DESCRIPTION
**Pull request by irevoire**
# Pull Request
To avoid complexifying the tmp_nodes stuff even more, I made the choice not to re-use the IDs freed **during** an indexing process, but to only re-use the IDs freed **between** different indexing processes.

## Related issue
Fixes https://github.com/meilisearch/arroy/issues/51

## What does this PR do?
- Add the capability to the concurrent nodeIDss to re-use available IDs without any mutexes or synchronizing anything

### CODE REVIEW

File: src/parallel.rs [hunk 1]
```diff
@@ -133,33 +134,90 @@ impl TmpNodesReader {
...
-    used: AtomicU64,
+    used: Arc<AtomicU64>,
+    /// A list of IDs to exhaust first before picking IDs from `current`.
+    /// When cloning this type, the available node ids are not cloned.
+    available: RoaringBitmap,
 }
 
 impl ConcurrentNodeIds {
     /// Creates the ID generator starting at the given number.
     pub fn new(used: RoaringBitmap) -> ConcurrentNodeIds {
```
Comment (Kerollmops) [inline src/parallel.rs:147]: Can you change the documentation?

File: src/parallel.rs [hunk 2]
```diff
@@ -133,33 +134,90 @@ impl TmpNodesReader {
...
 
-        ConcurrentNodeIds { current: AtomicU32::new(last_id), used: AtomicU64::new(used) }
+        ConcurrentNodeIds {
+            current: Arc::new(AtomicU32::new(last_id)),
+            used: Arc::new(AtomicU64::new(used_ids)),
+            available,
+        }
     }
 
     /// Returns and increment the ID you can use as a NodeId.
```
Comment (Kerollmops) [inline src/parallel.rs:159]: No more always incrementing, right?

File: src/parallel.rs [hunk 3]
```diff
@@ -133,33 +134,90 @@ impl TmpNodesReader {
...
+
+/// Cloning this type doesn't clone the available node ids
+impl Clone for ConcurrentNodeIds {
+    fn clone(&self) -> Self {
+        Self {
+            current: self.current.clone(),
+            used: self.used.clone(),
+            available: RoaringBitmap::new(),
+        }
+    }
```
Comment (Kerollmops) [inline src/parallel.rs:220]: Why? We shouldn't implement the `Clone` trait but rather a special cloning method.

File: src/writer.rs [hunk 4]
```diff
@@ -449,14 +462,19 @@ impl<D: Distance> Writer<D> {
...
                     &mut tmp_nodes,
                 )?;
                 assert!(node_id.mode != NodeMode::Item, "update_nodes_in_file returned an item even though there was more than a single element");
 
                 log::debug!("finished updating tree {root:X}");
-                Ok((node_id.unwrap_tree(), tmp_nodes.into_bytes_reader()?))
+                Ok(((node_id.unwrap_tree(), tmp_nodes.into_bytes_reader()?), concurrent_node_ids))
             })
-            .collect()
+            .collect::<Result<((Vec<ItemId>, Vec<TmpNodesReader>), Vec<ConcurrentNodeIds>)>>()?;
```
Comment (Kerollmops) [inline src/writer.rs:473]: Is it mandatory to specify the whole type, here?

File: src/writer.rs [hunk 5]
```diff
@@ -487,7 +506,7 @@ impl<D: Distance> Writer<D> {
                         Ok((NodeId::item(item_id), new_items))
                     }
                 } else if self.fit_in_descendant(new_items.len()) {
-                    let node_id = frozen_reader.concurrent_node_ids.next()?;
+                    let node_id = concurrent_node_ids.next()?;
                     let node_id = NodeId::tree(node_id as u32);
```
Comment (Kerollmops) [inline src/writer.rs:510]:
```suggestion
                    let node_id = NodeId::tree(node_id);
```

File: src/writer.rs [hunk 6]
```diff
@@ -637,36 +664,43 @@ impl<D: Distance> Writer<D> {
...
                     root_id.mode != NodeMode::Item,
                     "make_tree_in_file returned an item even though there was more than a single element"
                 );
                 log::debug!("finished generating tree {i:X}");
                 // make_tree will NEVER return a leaf when called as root
-                Ok((root_id.unwrap_tree(), tmp_nodes.into_bytes_reader()?))
+                Ok(((root_id.unwrap_tree(), tmp_nodes.into_bytes_reader()?), concurrent_node_ids))
             })
-            .collect()
+            .collect::<Result<((Vec<ItemId>, Vec<TmpNodesReader>), Vec<ConcurrentNodeIds>)>>()?;
```
Comment (Kerollmops) [inline src/writer.rs:699]: Do we need to specify the whole type, here?

File: src/writer.rs [hunk 7]
```diff
@@ -637,36 +664,43 @@ impl<D: Distance> Writer<D> {
         n_trees: Option<usize>,
         item_indices: &RoaringBitmap,
         frozen_reader: &FrozzenReader<D>,
-    ) -> Result<(Vec<ItemId>, Vec<TmpNodesReader>)> {
+        concurrent_node_ids: ConcurrentNodeIds,
+    ) -> Result<(Vec<ItemId>, Vec<TmpNodesReader>, ConcurrentNodeIds)> {
         let n_items = item_indices.len();
-        let concurrent_node_ids = frozen_reader.concurrent_node_ids;
+        let base_concurrent_node_id = concurrent_node_ids.clone();
```
Comment (Kerollmops) [inline src/writer.rs:670]: Why do you clone, here?

File: src/writer.rs [hunk 8]
```diff
@@ -637,36 +664,43 @@ impl<D: Distance> Writer<D> {
...
                 log::debug!("finished generating tree {i:X}");
                 // make_tree will NEVER return a leaf when called as root
-                Ok((root_id.unwrap_tree(), tmp_nodes.into_bytes_reader()?))
+                Ok(((root_id.unwrap_tree(), tmp_nodes.into_bytes_reader()?), concurrent_node_ids))
             })
-            .collect()
+            .collect::<Result<((Vec<ItemId>, Vec<TmpNodesReader>), Vec<ConcurrentNodeIds>)>>()?;
+
+        let concurrent_node_ids =
+            ConcurrentNodeIds::merge(concurrents_node_ids).unwrap_or(concurrent_node_ids);
```
Comment (Kerollmops) [inline src/writer.rs:702]: Can this unwrap happen?

File: src/writer.rs [hunk 9]
```diff
@@ -343,8 +341,15 @@ impl<D: Distance> Writer<D> {
                 n_items,
                 metadata.roots.len()
             );
-            let (new_roots, mut tmp_nodes_reader) =
-                self.update_trees(rng, metadata, &to_insert, &to_delete, &frozzen_reader)?;
+            let (new_roots, mut tmp_nodes_reader, new_concurrent_node_ids) = self.update_trees(
+                rng,
+                metadata,
+                &to_insert,
+                &to_delete,
```
Comment (Kerollmops) [inline src/writer.rs:348]: Can you avoid this warning?

**Review by Kerollmops — CHANGES_REQUESTED** [review]
Can we fix the _length comparison to zero_ warning, please? By using `!is_empty()`.

File: src/parallel.rs (hunk 3 above)
Comment (irevoire) [reply src/parallel.rs:220]: If I implement a special cloning method then I can't call `repeatn` later on :cry:

File: src/writer.rs (hunk 8 above)
Comment (irevoire) [reply src/writer.rs:702]: Yes, we don't know if we need to build more trees than what we have currently; thus, we may call `build_tree` for it to run 0 times

**Review by Kerollmops — APPROVED** [review]
It is much better with this `AtomicBool/AtomicU32` method!

**Pull request merged**